pkg/shared/accountmgmtutil/ams.go

```diff
@@ -68,144 +70,189 @@ func fetchOrgQuotaCost(ctx context.Context, conn connection.Connection) (*amscli
 
 }
 
-func GetUserSupportedInstanceTypes(ctx context.Context, spec *remote.AmsConfig, conn connection.Connection) (quota []QuotaSpec, err error) {
+func GetUserSupportedInstanceType(f *factory.Factory, spec *remote.AmsConfig, marketplace MarketplaceInfo) (*QuotaSpec, error) {
+
+	conn, err := f.Connection(connection.DefaultConfigSkipMasAuth)
+	if err != nil {
+		return nil, err
+	}
 
-	quotaCostGet, err := fetchOrgQuotaCost(ctx, conn)
+	quotaCostGet, err := fetchOrgQuotaCost(f.Context, conn)
 	if err != nil {
 		return nil, err
 	}
 
-	var quotas []QuotaSpec
+	var standardQuotas, marketplaceQuotas, trialQuotas []QuotaSpec
 	for _, quota := range quotaCostGet.GetItems() {
 		quotaResources := quota.GetRelatedResources()
 		for i := range quotaResources {
 			quotaResource := quotaResources[i]
 			if quotaResource.GetResourceName() == spec.ResourceName {
 				if quotaResource.GetProduct() == spec.TrialProductQuotaID {
-					quotas = append(quotas, QuotaSpec{QuotaTrialType, 0, quotaResource.BillingModel})
+					trialQuotas = append(trialQuotas, QuotaSpec{QuotaTrialType, 0, quotaResource.BillingModel, nil})
 				} else if quotaResource.GetProduct() == spec.InstanceQuotaID {
 					remainingQuota := int(quota.GetAllowed() - quota.GetConsumed())
-					quotas = append(quotas, QuotaSpec{QuotaStandardType, remainingQuota, quotaResource.BillingModel})
+					if quotaResource.BillingModel == "standard" {
+						standardQuotas = append(standardQuotas, QuotaSpec{QuotaStandardType, remainingQuota, quotaResource.BillingModel, nil})
+					} else if quotaResource.BillingModel == "marketplace" {
+						marketplaceQuotas = append(marketplaceQuotas, QuotaSpec{QuotaMarketplaceType, remainingQuota, quotaResource.BillingModel, quota.CloudAccounts})
+					}
 				}
 			}
 		}
 	}
 
-	return BattleOfInstanceBillingModels(quotas), err
-}
+	availableOrgQuotas := &OrgQuotas{standardQuotas, marketplaceQuotas, trialQuotas}
 
-// This function selects the billing model that should be used
-// It represents some requirement to always use the same standard billing models
-// This function should not exist but it does represents some requirement that we cannot do on backend
-func BattleOfInstanceBillingModels(quotas []QuotaSpec) []QuotaSpec {
-	var betterQuotasMap map[string]*QuotaSpec = make(map[string]*QuotaSpec)
-	alwaysWinsBillingModel := "standard"
-	for i := 0; i < len(quotas); i++ {
-		if quotas[i].BillingModel == alwaysWinsBillingModel {
-			betterQuotasMap[quotas[i].Name] = &quotas[i]
-		} else if betterQuotasMap[quotas[i].Name] == nil {
-			betterQuotasMap[quotas[i].Name] = &quotas[i]
+	return SelectQuotaForUser(f, availableOrgQuotas, marketplace)
+}
 
+func SelectQuotaForUser(f *factory.Factory, orgQuota *OrgQuotas, marketplaceInfo MarketplaceInfo) (*QuotaSpec, error) {
+	if len(orgQuota.StandardQuotas) == 0 && len(orgQuota.MarketplaceQuotas) == 0 {
+		if len(orgQuota.TrialQuotas) == 0 {
```

Comment:
Trial would be always present. Also doesn't need to checked in quota selection. 

 Decision we need to take here is: 
 - do we want to fail when user specify rubbish billing model or accoutID that doesn't exist? 
 - Can we build simple state machine for selection filtering that will be future proof?
 I have build diagram to showcase how this would look like:
 
<img width="1044" alt="Screenshot 2022-07-12 at 11 31 51" src="https://user-images.githubusercontent.com/981838/178470835-477a21fd-763e-47bc-b8be-84e0c8598f6d.png">